cmd/erasure-server-pool-rebalance.go

```diff
@@ -0,0 +1,935 @@
+	var wg sync.WaitGroup
+	for _, set := range pool.sets {
+		set := set
+		disks := set.getOnlineDisks()
+		if len(disks) == 0 {
+			logger.LogIf(ctx, fmt.Errorf("no online disks found for set with endpoints %s",
+				set.getEndpoints()))
+			continue
+		}
+
+		filterLifecycle := func(bucket, object string, fi FileInfo) bool {
+			if lc == nil {
+				return false
+			}
+			versioned := vc != nil && vc.Versioned(object)
+			objInfo := fi.ToObjectInfo(bucket, object, versioned)
+			action := evalActionFromLifecycle(ctx, *lc, lr, objInfo)
+			switch action {
+			case lifecycle.DeleteVersionAction, lifecycle.DeleteAction:
+				globalExpiryState.enqueueByDays(objInfo, false, action == lifecycle.DeleteVersionAction)
+				// Skip this entry.
+				return true
+			case lifecycle.DeleteRestoredAction, lifecycle.DeleteRestoredVersionAction:
+				globalExpiryState.enqueueByDays(objInfo, true, action == lifecycle.DeleteRestoredVersionAction)
+				// Skip this entry.
+				return true
+			}
+			return false
+		}
+
+		rebalanceEntry := func(entry metaCacheEntry) {
+			defer func() {
+				<-workers
+				wg.Done()
+			}()
+
+			if entry.isDir() {
+				return
+			}
+
+			// rebalance on poolIdx has reached its goal
+			if z.checkIfRebalanceDone(poolIdx) {
+				return
+			}
+
+			fivs, err := entry.fileInfoVersions(bucket)
+			if err != nil {
+				return
+			}
+
+			// We need a reversed order for rebalance,
+			// to create the appropriate stack.
+			versionsSorter(fivs.Versions).reverse()
+
+			var rebalanced int
+			for _, version := range fivs.Versions {
+				// Skip transitioned objects for now. TBD
+				if version.IsRemote() {
+					logger.LogIf(ctx, fmt.Errorf("skipping %s/%s transitioned object", bucket, version.Name))
+					continue
+				}
+
+				// Apply lifecycle rules on the objects that are expired.
+				if filterLifecycle(bucket, version.Name, version) {
+					logger.LogIf(ctx, fmt.Errorf("found %s/%s (%s) expired object based on ILM rules, skipping and scheduled for deletion", bucket, version.Name, version.VersionID))
+					continue
+				}
+
+				// We will skip rebalancing delete markers
+				// with single version, its as good as there
+				// is no data associated with the object.
+				if version.Deleted && len(fivs.Versions) == 1 {
+					logger.LogIf(ctx, fmt.Errorf("found %s/%s delete marked object with no other versions, skipping since there is no data to be rebalanced", bucket, version.Name))
+					continue
+				}
+
+				if version.Deleted {
+					_, err := z.DeleteObject(ctx,
+						bucket,
+						version.Name,
+						ObjectOptions{
+							Versioned:         vc.PrefixEnabled(version.Name),
+							VersionID:         version.VersionID,
+							MTime:             version.ModTime,
+							DeleteReplication: version.ReplicationState,
+							DeleteMarker:      true, // make sure we create a delete marker
+							SkipRebalancing:   true, // make sure we skip the decommissioned pool
+						})
+					var failure bool
+					if err != nil {
```

Comment:
Can we ignore object not found and version not found ? since the object can be deleted meanwhile